test/base.js

```diff
@@ -114,4 +116,69 @@ describe('base', () => {
     const result = m.predict(input);
     expect(result.error).to.be.an('error');
   });
-});
\ No newline at end of file
+
+  it('async predict', (done) => {
+    const m = xgb.XGModel('test/data/iris.xg.model');
+    const input = new Float32Array([
+      5.1, 3.5, 1.4, 0.2, // class 0
+      6.6, 3.0, 4.4, 1.4, // class 1
+      5.9, 3.0, 5.1, 1.8, // class 2
+    ]);
+    const mat = new xgb.matrix(input, 3, 4, NaN);
+    m.predictAsync(mat, 0, 0, (err, result) => {
+      // [
+      //   0, 1, 2,
+      //   3, 4, 5,
+      //   6, 7, 8,
+      // ]
+      expect(err).to.be.null;
+      expect(result[0] > 0.9).to.be.true; // class 0
+      expect(result[4] > 0.9).to.be.true; // class 1
+      expect(result[8] > 0.9).to.be.true; // class 2
+      done();
+    });
+  });
+
+  it('async parallel predict', function (done) {
+    this.timeout(120000);
+    const m = xgb.XGModel('test/data/iris.xg.model');
+    const input = [
+      { array: [5.1, 3.5, 1.4, 0.2], type: 0 }, // class 0
+      { array: [6.6, 3.0, 4.4, 1.4], type: 1 }, // class 1
+      { array: [5.9, 3.0, 5.1, 1.8], type: 2 }, // class 2
+    ];
+    let i = 0;
+    tasks = [];
+    function checkAsync(mat, mask, ntree, type, callback) {
+      mat = new xgb.matrix(mat, 1, 4, NaN);
+      m.predictAsync(mat, mask, ntree, (err, res) => {
+        expect(err).to.be.null;
+        expect(res[type] > 0.9);
+        callback();
+      });
+    }
+    while (i < 5000) {
+      input.forEach((v) => {
+        const task = async.apply(checkAsync,
+          new Float32Array(v.array), 0, 0, v.type
+        );
+        tasks.push(task);
+      });
+      i++;
+    }
+    async.parallelLimit(tasks, 1000, (err) => {
+      console.log('async.parallel tasks end');
+      expect(err).to.be.null;
+      done();
+    });
+    console.log('async.parallel tasks begin');
+  });
+
+  it('async error', function (done) {
```

Comment:
any other known error we can cover in testcases?

Reply:
Added restoreMatrix error test.